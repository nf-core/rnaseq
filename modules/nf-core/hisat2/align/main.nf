process HISAT2_ALIGN {
    tag "$meta.id"
    label 'process_high'

    // WARN: Version information not provided by tool on CLI. Please update version string below when bumping container versions.
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mulled-v2-a97e90b3b802d1da3d6958e0867610c718cb5eb1:2cdf6bf1e92acbeb9b2834b1c58754167173a410-0' :
        'biocontainers/mulled-v2-a97e90b3b802d1da3d6958e0867610c718cb5eb1:2cdf6bf1e92acbeb9b2834b1c58754167173a410-0' }"

    input:
    tuple val(meta), path(reads)
    tuple val(meta2), path(index)
    tuple val(meta3), path(splicesites)

    output:
    tuple val(meta), path("*.bam")                   , emit: bam
    tuple val(meta), path("*.log")                   , emit: summary
    tuple val(meta), path("*fastq.gz"), optional:true, emit: fastq
    path  "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def VERSION = '2.2.1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.

    def strandedness = ''
    if (meta.strandedness == 'forward') {
        strandedness = meta.single_end ? '--rna-strandness F' : '--rna-strandness FR'
    } else if (meta.strandedness == 'reverse') {
        strandedness = meta.single_end ? '--rna-strandness R' : '--rna-strandness RF'
    }
    ss = "$splicesites" ? "--known-splicesite-infile $splicesites" : ''
    def seq_center = params.seq_center ? "--rg-id ${prefix} --rg SM:$prefix --rg CN:${params.seq_center.replaceAll('\\s','_')}" : "--rg-id ${prefix} --rg SM:$prefix"
    if (meta.single_end) {
        def unaligned = params.save_unaligned ? "--un-gz ${prefix}.unmapped.fastq.gz" : ''
        """
        INDEX=`find -L ./ -name "*.1.ht2" | sed 's/\\.1.ht2\$//'`
        hisat2 \\
            -x \$INDEX \\
            -U $reads \\
            $strandedness \\
            $ss \\
            --summary-file ${prefix}.hisat2.summary.log \\
            --threads $task.cpus \\
            $seq_center \\
            $unaligned \\
            $args \\
            | samtools view -bS -F 4 -F 256 - > ${prefix}.bam

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            hisat2: $VERSION
            samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        END_VERSIONS
        """
    } else {
        def unaligned = params.save_unaligned ? "--un-conc-gz ${prefix}.unmapped.fastq.gz" : ''
        """
        INDEX=`find -L ./ -name "*.1.ht2" | sed 's/\\.1.ht2\$//'`
        hisat2 \\
            -x \$INDEX \\
            -1 ${reads[0]} \\
            -2 ${reads[1]} \\
            $strandedness \\
            $ss \\
            --summary-file ${prefix}.hisat2.summary.log \\
            --threads $task.cpus \\
            $seq_center \\
            $unaligned \\
            --no-mixed \\
            --no-discordant \\
            $args \\
            | samtools view -bS -F 4 -F 8 -F 256 - > ${prefix}.bam

        if [ -f ${prefix}.unmapped.fastq.1.gz ]; then
            mv ${prefix}.unmapped.fastq.1.gz ${prefix}.unmapped_1.fastq.gz
        fi
        if [ -f ${prefix}.unmapped.fastq.2.gz ]; then
            mv ${prefix}.unmapped.fastq.2.gz ${prefix}.unmapped_2.fastq.gz
        fi

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            hisat2: $VERSION
            samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        END_VERSIONS
        """
    }
}
