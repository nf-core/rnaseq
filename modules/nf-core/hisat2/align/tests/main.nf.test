nextflow_process {

    name "Test Process HISAT2_ALIGN"
    script "../main.nf"
    process "HISAT2_ALIGN"
    tag "modules"
    tag "modules_nfcore"
    tag "hisat2"
    tag "hisat2/align"
    tag "hisat2/build"
    tag "hisat2/extractsplicesites"

    test("Single-End") {

        setup {
            run("HISAT2_EXTRACTSPLICESITES") {
                script "../../extractsplicesites/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                """
                }
            }

            run("HISAT2_BUILD") {
                script "../../build/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.of([ [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                """
                }
            }
        }

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:true ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)]
                ])
                input[1] = HISAT2_BUILD.out.index
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.summary).match("se_summary") },
                { assert snapshot(process.out.fastq).match("se_fastq") },
                { assert snapshot(process.out.versions).match("se_versions") }
            )
        }
    }

    test("Paired-End") {

        setup {
            run("HISAT2_EXTRACTSPLICESITES") {
                script "../../extractsplicesites/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                """
                }
            }

            run("HISAT2_BUILD") {
                script "../../build/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.of([ [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                """
                }
            }
        }

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true) ]
                ])
                input[1] = HISAT2_BUILD.out.index
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.summary).match("pe_summary") },
                { assert snapshot(process.out.fastq).match("pe_fastq") },
                { assert snapshot(process.out.versions).match("pe_versions") }
            )
        }
    }

    test("Single-End No Splice Sites") {

        setup {
            run("HISAT2_BUILD") {
                script "../../build/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = [[:],[]]
                input[2] = [[:],[]]
                """
                }
            }
        }

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:true ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true) ]
                ])
                input[1] = HISAT2_BUILD.out.index
                input[2] = [[:],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.summary).match("se_no_ss_summary") },
                { assert snapshot(process.out.fastq).match("se_no_ss_fastq") },
                { assert snapshot(process.out.versions).match("se_no_ss_versions") }
            )
        }
    }

    test("Paired-End No Splice Sites") {

        setup {
            run("HISAT2_BUILD") {
                script "../../build/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = [[:],[]]
                input[2] = [[:],[]]
                """
                }
            }
        }

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true) ]
                ])
                input[1] = HISAT2_BUILD.out.index
                input[2] = [[:],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.summary).match("pe_no_ss_summary") },
                { assert snapshot(process.out.fastq).match("pe_no_ss_fastq") },
                { assert snapshot(process.out.versions).match("pe_no_ss_versions") }
            )
        }
    }
}
