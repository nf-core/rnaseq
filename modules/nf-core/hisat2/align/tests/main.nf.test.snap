{
    "se_versions": {
        "content": [
            [
                "versions.yml:md5,ceb638f44ebdaf09ba1f5c5c409585e2"
            ]
        ],
        "timestamp": "2023-10-16T15:14:50.269895296"
    },
    "se_no_ss_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.hisat2.summary.log:md5,7b8a9e61b7646da1089b041333c41a87"
                ]
            ]
        ],
        "timestamp": "2023-10-16T15:15:22.897386626"
    },
    "pe_no_ss_versions": {
        "content": [
            [
                "versions.yml:md5,ceb638f44ebdaf09ba1f5c5c409585e2"
            ]
        ],
        "timestamp": "2023-10-16T15:15:42.583699978"
    },
    "se_no_ss_versions": {
        "content": [
            [
                "versions.yml:md5,ceb638f44ebdaf09ba1f5c5c409585e2"
            ]
        ],
        "timestamp": "2023-10-16T15:15:22.909407356"
    },
    "pe_no_ss_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.hisat2.summary.log:md5,9839b31db795958cc4b70711a3414e9c"
                ]
            ]
        ],
        "timestamp": "2023-10-16T15:15:42.569775538"
    },
    "pe_no_ss_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-10-16T15:15:42.576881608"
    },
    "se_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.hisat2.summary.log:md5,7b8a9e61b7646da1089b041333c41a87"
                ]
            ]
        ],
        "timestamp": "2023-10-16T15:14:50.252466896"
    },
    "pe_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.hisat2.summary.log:md5,9839b31db795958cc4b70711a3414e9c"
                ]
            ]
        ],
        "timestamp": "2023-10-16T15:15:09.881690889"
    },
    "pe_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-10-16T15:15:09.888696129"
    },
    "se_no_ss_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-10-16T15:15:22.904010016"
    },
    "se_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-10-16T15:14:50.264366105"
    },
    "pe_versions": {
        "content": [
            [
                "versions.yml:md5,ceb638f44ebdaf09ba1f5c5c409585e2"
            ]
        ],
        "timestamp": "2023-10-16T15:15:09.894683308"
    }
}